controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -136,43 +180,88 @@ public void failed(State from, Throwable failure) {
+     *
+     * @param bucketIds set of bucket ids which need to be stopped.
+     * @param notify    if true will notify that abstract service stopped successfully or not.
+     */
+    public void stopBucketServices(Set<Integer> bucketIds, boolean notify) {
+        Futures.allOf(bucketIds.stream().map(bucketId -> stopBucketService(bucketId))
+                               .collect(Collectors.toList()))
+               .whenComplete((r, e) -> {
+                   if (e != null) {
+                       log.error("{}: bucket service shutdown failed with exception.", serviceType, e);
+                       if (notify) {
+                           notifyFailed(e);
+                       }
+                   } else {
+                       log.info("{}: bucket service stopped.", serviceType);
+                       if (notify) {
+                           notifyStopped();
+                       }
+                   }
+               });
+    }
+
+    private CompletableFuture<Void> stopBucketService(Integer bucketId) {
+        BucketService bucketService;
+        synchronized (lock) {
+            bucketService = buckets.get(bucketId);
+        }
+        CompletableFuture<Void> bucketFuture = new CompletableFuture<>();
+        bucketService.addListener(new Listener() {
+            @Override
+            public void terminated(State from) {
+                super.terminated(from);
+                synchronized (lock) {
+                    buckets.remove(bucketId);
+                    log.info("{}: Bucket service {} stopped.", serviceType, bucketId);
                 }
-            }, executor);
-            bucketService.stopAsync();
-
-            return bucketFuture;
-        }).collect(Collectors.toList()))
-                .whenComplete((r, e) -> {
-                    stopBucketOwnershipListener();
-                    if (e != null) {
-                        log.error("{}: bucket service shutdown failed with exception", serviceType, e);
-                        notifyFailed(e);
+                releaseBucketOwnership(bucketId, processId).thenCompose(released -> {
+                    if (!released) {
+                        log.warn("{}: Unable to stop bucket service  {} because it doesn't release it's ownership.",
+                                serviceType, bucketId);
+                        return startBucketService(bucketId);
+                    }
+                    return CompletableFuture.completedFuture(null);
+                }).whenComplete((r, e) -> {
+                    if (e == null) {
+                        bucketFuture.complete(null);
                     } else {
-                        log.info("{}: bucket service stopped", serviceType);
-                        notifyStopped();
+                        //if unable to stop the service, reacquire the ownership of bucket service.
+                        log.warn("{}: Reacquiring the ownership of bucket service {}.", serviceType, bucketId);
```

Comment:
Suppose a bucketService s1 is running on controller c1. After redistribution c1 need to release its ownership and another controller instance need to own it. But due some release c1 is facing issue to release it ownership. In that case, instead of keeping the bucket service in faulty state, controller will reinitiate the service and reacquire the ownership of bucket.

Reply:
But this means that the Controller is owning a bucket that, according to the new assignment of buckets to controller, it should not be owning, right? If problems to shutdown that specific bucket continue, the bucket may not be owned by anyone else, right? I don't know if this has any simple solution, but just want to understand the decisions being taken in the PR.

Reply:
> But this means that the Controller is owning a bucket that, according to the new assignment of buckets to controller, it should not be owning, right?

Yes, that's right.

> but just want to understand the decisions being taken in the PR

To keep actual number of services running same as number of  bucket service configured, we implement this  reacquire ownership functionality. Suppose we defined total number of bucket service as 10 among 2 instance of controller. After redistribution suppose c1 need to release one bucket and c2 need to own it. But due to any reason if c1 unable to release the ownership. Until and unless C1 will not release, C2 couldn't own it. If this logic (reacquire ownership) is not there, then it might be the case that we configured total no of bucket services as 10 but actual running are 9 only.